### PR TITLE
Add Flatcar Container Linux template

### PR DESCRIPTION
**Pull request by jepio**
**What type of PR is this?**
/kind other

**What this PR does / why we need it**:

This PR adds a cluster template for Flatcar Container Linux based clusters.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Related to https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/629

**Special notes for your reviewer**:
The Flatcar cluster template works with images built using this image-builder PR https://github.com/kubernetes-sigs/image-builder/pull/701 and this cluster-api PR https://github.com/kubernetes-sigs/cluster-api/pull/4172.

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

No image versions are changed.

**TODOs**:

- [X] squashed commits <- *single commit*
- [ ] includes documentation 
- [ ] adds unit tests

**Release note**:

```release-note
Flatcar Container Linux is now supported as a cluster flavor. Try `clusterctl generate cluster my-cluster --flavor flatcar`.
```

### CODE REVIEW

File: templates/flavors/flatcar/machine-deployment.yaml
Comment (invidian) [inline templates/flavors/flatcar/machine-deployment.yaml:46]: It seems those values (quotes) and order of keys could be consistent with other files.

**Review by invidian — COMMENTED** [review]
Just nits, overall looks good!

File: templates/flavors/flatcar/machine-deployment.yaml
Comment (jepio) [reply templates/flavors/flatcar/machine-deployment.yaml:46]: I tried to order these field in (what I considered) logical order:
* subscription -> resource group -> gallery -> (image) name
* plan publisher -> offer ->sku
but the envsubst step seems to reorder the fields alphabetically. So i'll reorder this.

Are you suggesting removing the quotes?

File: templates/flavors/flatcar/machine-deployment.yaml
Comment (jepio) [reply templates/flavors/flatcar/machine-deployment.yaml:73]: Sure

File: templates/flavors/flatcar/machine-deployment.yaml
Comment (invidian) [reply templates/flavors/flatcar/machine-deployment.yaml:46]: Between the files, I see:
- subscriptionID, resourceGroup, gallery, name, ...
- subscriptionID, resourceGroup, name, gallery, ...
- gallery, name, resourceGroup, sku, subscriptionID, ...

>but the envsubst step seems to reorder the fields alphabetically. So i'll reorder this.

I remember I was fighting with this as well I think it's a good idea to have them ordered as after the rendering, as then it's possible to diff between the template and actual configs (which in turn might be useful for updating your configuration).

File: templates/flavors/flatcar/patches/kubeadm-controlplane.yaml
Comment (invidian) [inline templates/flavors/flatcar/patches/kubeadm-controlplane.yaml:61]: Missing newline at the end of the file. It's nice to have it, so when you do `cat ...`, it does not mess up your prompt :)

File: templates/flavors/flatcar/machine-deployment.yaml
Comment (jepio) [reply templates/flavors/flatcar/machine-deployment.yaml]: The fields are now in the same order as in the rendered template and quotes are gone wherever the rendered template removes them.

File: templates/flavors/flatcar/machine-deployment.yaml
Comment (jepio) [reply templates/flavors/flatcar/machine-deployment.yaml]: Gone

File: templates/flavors/flatcar/patches/kubeadm-controlplane.yaml
Comment (jepio) [reply templates/flavors/flatcar/patches/kubeadm-controlplane.yaml]: Should be there now

File: templates/flavors/flatcar/machine-deployment.yaml
Comment (invidian) [reply templates/flavors/flatcar/machine-deployment.yaml]: Awesome, thank you!

**Review by invidian — APPROVED** [review]

File: templates/flavors/flatcar/patches/kubeadm-controlplane.yaml
Comment (CecileRobertMichon) [inline templates/flavors/flatcar/patches/kubeadm-controlplane.yaml:62]: Does this require the user to build their own image in a SIG first? if so, can we put some docs around it? Also, where is this version coming from?

Comment (jepio) [reply templates/flavors/flatcar/patches/kubeadm-controlplane.yaml:62]: The version contains the build timestamp and comes from here: https://github.com/kubernetes-sigs/image-builder/blob/master/images/capi/packer/azure/azure-sig.json.

At this time: yes, the user has to build their own images. There is existing documentation on building SIG images here https://capz.sigs.k8s.io/topics/custom-images.html#using-shared-image-gallery-recommended, and it even mentions Flatcar already :) What extensions to the documentation would you suggest?

Moving forward I'd love to see new marketplace images published for testing/demo purposes. I'd love to hear your thoughts on that. We could either publish third-party images under the existing "Kinvolk" publisher, or perhaps first-party is an option?

Comment (jepio) [reply templates/flavors/flatcar/patches/kubeadm-controlplane.yaml:62]: I think I found a better way:
```
version: latest
```
is also correct.

**Comment by invidian** [comment]
>Does this require kubernetes-sigs/cluster-api#4172 to merge first?

It does, yes, as it's using `ignition` API field.

**Comment by CecileRobertMichon** [comment]
/hold

until kubernetes-sigs/cluster-api#4172 is released

**Comment by invidian** [comment]
CABPK PR kubernetes-sigs/cluster-api#4172 has been merged and will be part of v1.1 release, so we're getting closer getting this as well :tada:

**Comment by jepio** [comment]
Updated the apiversion from `alpha4` to `beta1`, and rebased. I also tested that the template continues to work with the ignition support merged to CAPI and it does 👍

**Comment by invidian** [comment]
I wonder, CAPI v1.1 RC release should be happening soon. I wonder if we can update CAPA main manifests to use it, so this PR could be merged.

**Comment by CecileRobertMichon** [comment]
> I wonder, CAPI v1.1 RC release should be happening soon. I wonder if we can update CAPA main manifests to use it, so this PR could be merged.

Do you mean CAPZ? We are about to release CAPZ v1.1 so I don't want to merge this until CAPI release happens (in January) so we don't end up releasing a flavor that doesn't work.

**Comment by invidian** [comment]
I was thinking about updating https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/d352e7c6ea47974d8ec7c858cb76dea7e9d3ecc3/test/e2e/config/azure-dev.yaml#L60-L61 to CAPBK v1.1 rc1 release as an intermediate step before CAPBK v1.1 gets released, so we could merge this PR then.

**Comment by CecileRobertMichon** [comment]
#1951 just merged which means this should now work 

how do we test this template to make sure it doesn't regress?

File: templates/flavors/flatcar/patches/kubeadm-controlplane.yaml
Comment (CecileRobertMichon) [inline templates/flavors/flatcar/patches/kubeadm-controlplane.yaml:62]: @jsturtevant FYI this is a similar issue to what you were encountering in #1729. This template requires an image in a shared gallery since there is currently no published flatcar image in Azure.

@invidian @jepio what are the steps for someone to use this from their own subscription? Do they need to first build an image?

Comment (invidian) [reply templates/flavors/flatcar/patches/kubeadm-controlplane.yaml:62]: >@invidian @jepio what are the steps for someone to use this from their own subscription? Do they need to first build an image?

This is similar to CAPA. Right now we build and publish our own AMIs, but soon they will be build by the CAPA team, as Flatcar's support is included in upstream image-builder.

In general, from what we know about CAPI, users should be responsible for building their own images and images which are already published are intended to be only for testing.

How is this handled for the images for e2e tests which are already used by CAPZ?

**Comment by invidian** [comment]
>how do we test this template to make sure it doesn't regress?

Hmm, so we're open to add e2e tests for Flatcar if you already have e2e tests for other flavors. But we need to sort out the issue with images first I guess.

File: templates/flavors/flatcar/patches/kubeadm-controlplane.yaml
Comment (jepio) [reply templates/flavors/flatcar/patches/kubeadm-controlplane.yaml:62]: So far we've been testing manually. This requires building an image in your subscription:
```
cd image-builder/images/capi
make build-azure-sig-flatcar FLATCAR_VERSION=current
```
Then the gallery image parameters need to be set in the template that this PR adds, and you're all set.

**Comment by invidian** [comment]
>Hmm, so we're open to add e2e tests for Flatcar if you already have e2e tests for other flavors. But we need to sort out the issue with images first I guess.

I've just pushed e2e tests configuration for Flatcar, which I can use locally to deploy a cluster. However, I used `KUBERNETES_VERSION=v1.21.9` and `AZURE_LOCATION=uksouth` for generating workload cluster configuration and this is how SIG is configured, so e2e tests may not pass on CI.

So I guess once Flatcar images are built in some SIG we could safely use on CI, we are good to go.

**Comment by invidian** [comment]
e2e tests passes, but I guess Flatcar cluster is not created then? Maybe some wiring is missing? I was using `kcp-scale-in` e2e test flavor as a reference.

**Comment by CecileRobertMichon** [comment]
> So I guess once Flatcar images are built in some SIG we could safely use on CI, we are good to go.

SIG images are only accessible by the subscription that created them. This means the tests wouldn't able to run locally for developers testing with their own subscription.

**Comment by invidian** [comment]
>I don't see an e2e test added to the test suite. See https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/test/e2e/azure_test.go#L268 for an example.

Thanks. I've added test for Flatcar flavor and it's passing for me locally.

I also spotted this in the test logs, I'll have a closer look into it:
`[2] Failed to get logs for machine capz-e2e-uursk6-flatcar-control-plane-4bq5f, cluster capz-e2e-uursk6/capz-e2e-uursk6-flatcar: [running command "journalctl --no-pager --output=short-precise": Process exited with status 1, running command "journalctl --no-pager --output=short-precise -u containerd.service": Process exited with status 1, running command "cat /var/log/cloud-init.log": Process exited with status 1, running command "kubelet --version": Process exited with status 127, running command "journalctl --no-pager --output=short-precise -u kubelet.service": Process exited with status 1, running command "cat /var/log/cloud-init-output.log": Process exited with status 1, running command "journalctl --no-pager --output=short-precise -k": Process exited with status 1]`

File: Makefile
Comment (CecileRobertMichon) [inline Makefile:452]: instead of doing this, can we add a "test" flavor? See https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/templates/test/ci/prow-ipv6/kustomization.yaml for an example of how to do this.

The templates in infrastructure-azure/v1beta1 are only for the CAPI test suite. The Azure specific flavors for prow jobs are kept in templates/test/ci. They are based on the regular flavors with an added CRS for CNI and some resource tags to make sure the resources are cleaned up in the test sub if the test is interrupted.

Comment (invidian) [reply Makefile]: Done, I think, though I'm not sure how I can test this locally, so I just reverted this change and pushed a new one for now. Hopefully CI fails successfully again :smile:

Comment (invidian) [reply Makefile]: So the test seems to be skipped right now. Is that desired?

File: test/e2e/azure_test.go
Comment (CecileRobertMichon) [inline test/e2e/azure_test.go:270]: please add a Context around this It Block 

Then we should mark it as either [REQUIRED] or [OPTIONAL]. [REQUIRED] will run it make it run on every PR and require pass for merge, while [OPTIONAL] will only run it on PRs when the optional job is triggered and will run periodically on main. I think we can start with [OPTIONAL] for now, wdyt?

Comment (invidian) [reply test/e2e/azure_test.go]: Sure, `[OPTIONAL]` sounds good. I don't have experience with Ginkgo or your naming convention for tests, so I'm not sure if `Context` and `It` blocks are properly named. I'll adjust accordingly if you have some suggestions.

File: test/e2e/azure_test.go
Comment (CecileRobertMichon) [inline test/e2e/azure_test.go:281]: flavor needs to be defined in fake github release in azure-dev.yaml

https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1729/files#diff-35f348a5d567ebeba39b1e40bbab0e1a7f64009cbadff142952f9b9691e94666R136

Comment (invidian) [reply test/e2e/azure_test.go:281]: Done IIUC.

File: templates/flavors/flatcar/machine-deployment.yaml
Comment (jepio) [reply templates/flavors/flatcar/machine-deployment.yaml:35]: The current thing is addressed by this PR: https://github.com/kubernetes-sigs/image-builder/pull/834/

File: azure/services/resourceskus/client.go
Comment (invidian) [inline azure/services/resourceskus/client.go:60]: Would be good to doble check if it's fine to put `"true"` in here.

**Comment by invidian** [comment]
Flatcar e2e test fails because the community gallery is in the wrong location. I'll fix that next week.

**Comment by invidian** [comment]
/test pull-cluster-api-provider-azure-e2e-optional

Replicated image to `eastus2`, let's see if test passes now :crossed_fingers:

**Comment by invidian** [comment]
e2e tests for Flatcar passed in https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/1729/pull-cluster-api-provider-azure-e2e/1521131029673283584.

**Comment by invidian** [comment]
/hold

Until community gallery references are updated to something more reasonable than my test community gallery.

@CecileRobertMichon e2e tests for Flatcar are now passing using community gallery, so this PR can be reviewed again, but shouldn't be merged as mentioned above.

**Comment by invidian** [comment]
In case community galleries support gets merged with this PR, we might also modify PR description before merging to mention it resolves https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/2267. I don't have permissions to do so.

**Comment by CecileRobertMichon** [comment]
@invidian would it be easier/better to extract the community gallery support to a separate PR? If so, feel free to do so.

**Comment by invidian** [comment]
>@invidian would it be easier/better to extract the community gallery support to a separate PR? If so, feel free to do so.

Here the community gallery has e2e tests with it, but on the other hand this PR may still take a bit before merging, as we need to do final cleanups with the used images. Will open separate PR.

**Comment by invidian** [comment]
Opened https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2277 then.

File: azure/services/resourceskus/client.go
Comment (invidian) [reply azure/services/resourceskus/client.go:60]: Can be resolved, will be discussed in https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2277.

**Comment by invidian** [comment]
I will be working on https://github.com/flatcar-linux/Flatcar/issues/734 to get the community gallery out with some images we can use here. I think that's the last requirement missing for this PR :sweat_smile:

**Comment by invidian** [comment]
/retest

Flatcar test has failed as I forgot to replicate the image to all regions used by CI. Should be fixed now.

**Comment by invidian** [comment]
/retest

Flatcar tests are passing again, but let's try making it green.

**Comment by invidian** [comment]
/test pull-cluster-api-provider-azure-e2e-optional

Trying out using gen 2 images but I hit some issues with wrong SKU and missing replication region for an image. Hopefully it will pass now.

**Comment by invidian** [comment]
https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2890 is now merged :)

**Pull request closed without merging**
